src/OpenTelemetry/SuppressInstrumentation.cs

```diff
@@ -0,0 +1,77 @@
+﻿// <copyright file="SuppressInstrumentation.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using OpenTelemetry.Context;
+
+namespace OpenTelemetry
+{
+    public class SuppressInstrumentation : IDisposable
+    {
+        private static readonly RuntimeContextSlot<SuppressInstrumentation> SuppressInstrumentationRuntimeContextSlot = RuntimeContext.RegisterSlot<SuppressInstrumentation>("otel.suppress_instrumentation");
+
+        private readonly SuppressInstrumentation parent;
+        private bool disposed;
+
+        private SuppressInstrumentation()
+        {
+            this.parent = SuppressInstrumentationRuntimeContextSlot.Get();
+            SuppressInstrumentationRuntimeContextSlot.Set(this);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether automatic telemetry
+        /// collection in the current context should be suppressed (disabled).
+        /// </summary>
+        public static bool IsSuppressed => SuppressInstrumentationRuntimeContextSlot.Get() != default;
```

Comment:
Consider implicate operator?

```csharp
public sealed class SuppressInstrumentationScope : IDisposable
{
    private readonly bool value;
    public static implicit operator bool(SuppressInstrumentationScope _) => SuppressInstrumentationRuntimeContextSlot.Get();

    public SuppressInstrumentationScope(bool value = true)
    {
        this.value = value;
        SuppressInstrumentationRuntimeContextSlot.Set(value);
    }

    public void Dispose()
    {
        SuppressInstrumentationRuntimeContextSlot.Set(this.value);
    }

    public IDisposable Begin(bool value = true)
    {
        return new SuppressInstrumentationScope(value);
    }
}

public static SuppressInstrumentationScope SuppressInstrumentation { get; set; } = new SuppressInstrumentationScope(false);
```

Reply:
The for consumption:

```csharp
if (Sdk.SuppressInstrumentation)
{
    // do something
}

using (Sdk.SuppressInstrumentation.Begin(true))
{
    // do something
}
```

Reply:
Interesting, I would not have thought of this approach!

Reply:
Hmm, this isn't working for some reason. I'll take closer look in the morning.

Basically, I debugged through this code and...
```csharp
using (SuppressInstrumentationScope.Begin()) // This does in fact invoke SuppressInstrumentationRuntimeContextSlot.Set(true);
{
    Assert.True(Sdk.SuppressInstrumentation); // This fails, even though the implicit operator is invoked
}
```

In the meantime I had a few changes staged prior to seeing this idea, so I just pushed them up.

Reply:
Oh.. silly me, the static initializer for `Sdk.SuppressInstrumentation` had not yet been invoked prior to calling `SuppressInstrumentationScope.Begin()`

Reply:
Ok pushed up something that works. Ignore my initial confused comments.